### PR TITLE
Fixes #848 add CLI mode to manually concatenate ts parts

### LINKED ISSUE #848: [Feature request] TwitchDownloaderCLI: add option to merge existing ts parts into custom output.ts
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues?q=is%3Aissue) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Write your feature request here

What I am requesting is a way to directly call this method:
https://github.com/lay295/TwitchDownloader/blob/fefd3d18f84185977b9959032dd423b7e1fed405/TwitchDownloaderCore/VideoDownloader.cs#L663-L698

so I can merge already downloaded ts parts into custom `output.ts`, for example using ranges (like ffmpeg does) and individual files:
```
TwitchDownloaderCLI merge 10.ts -ss 14-muted.ts -to 17-muted.ts 20.ts -o output_custom.ts
```
So `custom.ts` will be made of (in the same order as given in the argument):
```
10.ts
14-muted.ts
15-muted.ts
16-muted.ts
17-muted.ts
20.ts
```

This custom way of calling CombineVideoParts call can help debug issues merging ts files with .NET Framework.

Also please add all the information to documentation.

### PR DESCRIPTION
**Pull request by superbonaci**
Fix https://github.com/lay295/TwitchDownloader/issues/848
There's a weird github issue which looks like 2 lines were modified:
https://github.com/orgs/community/discussions/72813

### CODE REVIEW

**Comment by superbonaci** [comment]
Looks like visual Studio for Mac introduces weird ends of line and messes up the BOM in files:
https://github.com/microsoft/vscode/issues/196905

**Comment by superbonaci** [comment]
Files have to be created with Visual Studio for Windows, so will open another pull request once it's fixed.

**Comment by superbonaci** [comment]
Looks like know is kind of fixed.

**Comment by superbonaci** [comment]
Some of the original .cs files end in newline and other don't, so they are mixed up.

File: TwitchDownloaderCLI/Program.cs [hunk 1]
```diff
@@ -24,13 +24,14 @@ private static void Main(string[] args)
...
 
-            var parserResult = parser.ParseArguments<VideoDownloadArgs, ClipDownloadArgs, ChatDownloadArgs, ChatUpdateArgs, ChatRenderArgs, FfmpegArgs, CacheArgs>(preParsedArgs);
+            var parserResult = parser.ParseArguments<TsMergeArgs, VideoDownloadArgs, ClipDownloadArgs, ChatDownloadArgs, ChatUpdateArgs, ChatRenderArgs, FfmpegArgs, CacheArgs>(preParsedArgs);
             parserResult.WithNotParsed(errors => WriteHelpText(errors, parserResult, parser.Settings));
 
             CoreLicensor.EnsureFilesExist(AppContext.BaseDirectory);
             WriteApplicationBanner((ITwitchDownloaderArgs)parserResult.Value, args);
 
             parserResult
+                .WithParsed<TsMergeArgs>(MergeTS.Merge)
```
Comment (ScrubN) [inline TwitchDownloaderCLI/Program.cs:34]: Why is it placed at the top? The previous order was more or less in order of importance.

File: TwitchDownloaderCLI/README.md [hunk 2]
```diff
@@ -1,7 +1,9 @@
 # TwitchDownloaderCLI
 A cross platform command line tool that can do the main functions of the GUI program, which can download VODs/Clips/Chats and render chats.
+Also can concatenate/combine/merge Transport Stream files, either those parts downloaded with the CLI itself or from another source.
 
 - [TwitchDownloaderCLI](#twitchdownloadercli)
+  - [Arguments for mode tsmerge](#arguments-for-mode-tsmerge)
```
Comment (ScrubN) [inline TwitchDownloaderCLI/README.md:6]: Again, it's at the top

File: TwitchDownloaderCLI/README.md [hunk 3]
```diff
@@ -14,6 +16,18 @@ A cross platform command line tool that can do the main functions of the GUI pro
...
+
+**-l / --inputlist (REQUIRED)**
+Path to input list file in text format (one part per line).
+
+**-o / --output (REQUIRED)**
+File the program will output to.
+
+**--banner**
+(Default: `true`) Displays a banner containing version and copyright information.
+
```
Comment (ScrubN) [inline TwitchDownloaderCLI/README.md:30]: Top again

File: TwitchDownloaderCLI/README.md [hunk 4]
```diff
@@ -348,6 +362,10 @@ Other = `1`, Broadcaster = `2`, Moderator = `4`, VIP = `8`, Subscriber = `16`, P
 ## Example Commands
 <sup>Examples of typical use cases</sup>
 
+Concatenate several ts parts into another, keeping all streams well formatted
+
+    TwitchDownloaderCLI tsmerge -l list.txt -o output.ts
+
```
Comment (ScrubN) [inline TwitchDownloaderCLI/README.md:368]: Top again

File: TwitchDownloaderCore/TsMerger.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+                _progress.Report(new ProgressReport() { ReportType = ReportType.NewLineStatus, Data = "Combining Parts 0% [2/2]" });
+
+                await CombineVideoParts(downloadFolder, videoPartsList, cancellationToken);
+
+                _progress.Report(new ProgressReport(100));
+            }
+            finally
+            {
+            }
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:45]: this try-finally is redundant

File: TwitchDownloaderCore/TsMerger.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+            private void VerifyDownloadedParts(List<string> videoParts, string downloadFolder, CancellationToken cancellationToken)
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:48]: Nitpick: indentation

File: TwitchDownloaderCLI/Modes/Arguments/TsMergeArgs.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using CommandLine;
+
+namespace TwitchDownloaderCLI.Modes.Arguments
+{
+    [Verb("tsmerge", HelpText = "Concatenates (not binary) .ts/.tsv/.tsa/.m2t/.m2ts (MPEG Transport Stream) parts into another file")]
```
Comment (ScrubN) [inline TwitchDownloaderCLI/Modes/Arguments/TsMergeArgs.cs:5]: What does "Concatenates (not binary)" mean? This is not made clear to the user.

File: TwitchDownloaderCLI/Modes/Arguments/TsMergeArgs.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    public class TsMergeArgs : ITwitchDownloaderArgs
+    {
+        [Option('l', "inputlist", Required = true, HelpText = "Path to input list file in text format (one part per line).")]
```
Comment (ScrubN) [inline TwitchDownloaderCLI/Modes/Arguments/TsMergeArgs.cs:8]: This is wording not very clear.

File: TwitchDownloaderCore/TsMerger.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+        public TsMerger(TsMergeOptions tsMergeOptions, IProgress<ProgressReport> progress)
+        {
+            downloadOptions = tsMergeOptions;
+            _progress = progress;
+        }
+
+        public async Task MergeAsync(CancellationToken cancellationToken)
+        {
+            string downloadFolder = "";
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:25]: `downloadFolder` is never reassigned?

File: TwitchDownloaderCore/TsMerger.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+            if (failedParts.Count != 0)
+            {
+                if (failedParts.Count == videoParts.Count)
+                {
+                    // Every video part returned corrupted, probably a false positive.
+                    return;
+                }
+
+                _progress.Report(new ProgressReport(ReportType.Log, $"The following parts appear to be invalid TS files: {string.Join(", ", failedParts)}"));
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:77]: This does not discriminate between missing and corrupted files.

**Review by ScrubN — CHANGES_REQUESTED** [review]
I don't like how you had to copy-paste functions from the video downloader, however there really isn't another way until kick-support is finished and merged.

Don't get me wrong, I see value in this function and have been waiting to add it myself, but I would like tsmerge to be moved below the primary 5 functions. I do see it being used by some power users, but not nearly enough to warrant being placed above the other functions in the help text & readme.

**Comment by superbonaci** [comment]
I'll make some adjustments to it. What do you suggest instead of this line?
```
_progress.Report(new ProgressReport(ReportType.Log, $"The following parts appear to be invalid TS files: {string.Join(", ", failedParts)}"));
```

**Comment by ScrubN** [comment]
> I'll make some adjustments to it. What do you suggest instead of this line?
> 
> ```
> _progress.Report(new ProgressReport(ReportType.Log, $"The following parts appear to be invalid TS files: {string.Join(", ", failedParts)}"));
> ```

Instead of dealing with discrimination, maybe remove the file exist check and let the exception propagate up. Since this is not an automatic process, the user is liable for the files existing.

File: TwitchDownloaderCore/TsMerger.cs [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+        private async Task CombineVideoParts(string downloadFolder, List<string> videoParts, CancellationToken cancellationToken)
+        {
+            DriveInfo outputDrive = DriveHelper.GetOutputDrive(downloadFolder);
+            string outputFile = Path.Combine(downloadFolder, downloadOptions.OutputFile);
+
+            int partCount = videoParts.Count;
+            int doneCount = 0;
+
+            await using var outputStream = new FileStream(outputFile, FileMode.Create, FileAccess.Write, FileShare.None);
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:109]:
```suggestion
            await using var outputStream = new FileStream(outputFile, FileMode.Create, FileAccess.Write, FileShare.Read);
```

File: TwitchDownloaderCore/TsMerger.cs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            foreach (var part in videoParts)
+            {
+                await DriveHelper.WaitForDrive(outputDrive, _progress, cancellationToken);
+
+                string partFile = Path.Combine(downloadFolder, part);
+                if (File.Exists(partFile))
+                {
+                    await using (var fs = File.Open(partFile, FileMode.Open, FileAccess.Read, FileShare.None))
```
Comment (ScrubN) [inline TwitchDownloaderCore/TsMerger.cs:117]:
```suggestion
                    await using (var fs = File.Open(partFile, FileMode.Open, FileAccess.Read, FileShare.Read))
```

**Review by ScrubN — CHANGES_REQUESTED** [review]

**Comment by superbonaci** [comment]
> tsmerge to be moved below the primary 5 functions. I do see it being used by some power users, but not nearly enough to warrant being placed above the other functions in the help text & readme.

Concatenating ts parts is certainly a corner case, and for that reason most people rely on tools (like ffmpeg) that at least do something, but they don't achieve a good result. The method implemented by .NET Framework is ideal at least with the samples I've tested. So don't think it's not useful.

I'll made a new pull request with the changes you suggested, hope now they are ok.

**Pull request closed without merging**